rust/src/transformations/make_stable_expr/expr_fill_null/mod.rs

```diff
@@ -0,0 +1,153 @@
+
+    let t_prior = input
+        .clone()
+        .make_stable(input_domain.clone(), input_metric.clone())?;
+    let (middle_domain, middle_metric) = t_prior.output_space();
+
+    let mut output_domain = middle_domain.clone();
+    let series_domain = output_domain.active_series_mut()?;
+
+    use DataType::*;
+    let constant = match &series_domain.field.dtype {
+        UInt32 => reconcile_domain::<u32>(constant, series_domain),
+        UInt64 => reconcile_domain::<u64>(constant, series_domain),
+        Int8 => reconcile_domain::<i8>(constant, series_domain),
+        Int16 => reconcile_domain::<i16>(constant, series_domain),
+        Int32 => reconcile_domain::<i32>(constant, series_domain),
+        Int64 => reconcile_domain::<i64>(constant, series_domain),
+        Float32 => reconcile_domain::<f32>(constant, series_domain),
+        Float64 => reconcile_domain::<f64>(constant, series_domain),
+        Boolean => reconcile_domain::<bool>(constant, series_domain),
+        String => reconcile_domain::<str>(constant, series_domain),
+        UInt8 | UInt16 => {
+            fallible!(
+                MakeTransformation,
+                "u8 and 16 are not supported, please use u32 or u64 instead"
+            )
+        }
+        dtype => fallible!(MakeTransformation, "unsupported dtype: {}", dtype),
+    }?;
+
+    t_prior
+        >> Transformation::new(
+            middle_domain.clone(),
+            output_domain,
+            Function::then_expr(move |expr| expr.fill_null(constant.clone())),
+            middle_metric.clone(),
+            middle_metric,
+            StabilityMap::new(Clone::clone),
+        )?
+}
+
+pub fn match_fill_null(expr: &Expr) -> Fallible<Option<(&Expr, &LiteralValue)>> {
+    let Expr::Function {
+        input,
+        function: FunctionExpr::FillNull { .. },
+        ..
+    } = expr
+    else {
+        return Ok(None);
+    };
+
+    let Ok([input, constant]) = <&[_; 2]>::try_from(input.as_slice()) else {
+        return fallible!(MakeTransformation, "fill_null expects 2 arguments");
+    };
+
+    let Expr::Literal(constant) = constant else {
+        return fallible!(
+            MakeTransformation,
+            "fill_null expects a literal as second argument"
+        );
+    };
+    Ok(Some((input, constant)))
+}
+
+fn reconcile_domain<T: TryGetValue + ?Sized>(
+    constant: &LiteralValue,
+    series_domain: &mut SeriesDomain,
+) -> Fallible<Expr>
+where
+    T::Owned: Primitive + Literal,
+{
+    let constant = constant
+        .to_any_value()
+        .ok_or_else(|| err!(MakeTransformation))?
+        .strict_cast(&series_domain.field.dtype)?;
+    let constant = T::try_get_value(constant)?;
+
+    if !series_domain.atom_domain::<T::Owned>()?.member(&constant)? {
+        return fallible!(
+            MakeTransformation,
+            "constant {:?} is not in domain {:?}",
+            constant,
+            series_domain
+        );
+    }
+
+    series_domain.nullable = false;
```

Comment:
I'm sure this has a role, and the signature obviously marks it as mutable, but it still feels like a side effect. Would it make any sense to pull this out, and the function would only concern itself with checking and rewrapping `constant`?

Reply:
Ah yes, this case is easy to pull out because it's not mutating a typed value.